### ISSUE TITLE
Issue with using Symfony Serializer

### ISSUE THREAD

**Issue by aik099** [acmephp/acmephp#308]
When attempting to regenerate SSL certificate after 3 months I'm getting these errors both on master (PHP 8.3) and latest release (PHP 7.4):

```
Registering contact my@email.com...
Loading account key pair...  
Registering on the ACME server...
Account registered successfully!

In Repository.php line 238:
                                                            
  Loading of domain my.domain.com distinguished name failed  
                                                            

In Serializer.php line 224:
                                                                                                         
  Could not denormalize object of type "AcmePhp\Ssl\DistinguishedName", no supporting normalizer found.  
                                                                                                         

run [-d|--delay DELAY] [--] <config>
```

The issue is with dependencies. If I stay on 2.1.0 with these dependency versions (`composer show -i` command), then all works:

```
adbario/php-dot-notation          2.2.0   PHP dot notation access to arrays
alibabacloud/cdn                  1.8.958 Alibaba Cloud Cdn SDK for PHP
alibabacloud/client               1.5.31  Alibaba Cloud Client for PHP - Use Alibaba Cloud in your PHP project
alibabacloud/wafopenapi           1.8.884 Alibaba Cloud WafOpenapi SDK for PHP
aws/aws-crt-php                   v1.0.2  AWS Common Runtime for PHP
aws/aws-sdk-php                   3.198.6 AWS SDK for PHP - Use Amazon Web Services in your PHP project
clagiordano/weblibs-configmanager v1.2.0  weblibs-configmanager is a tool library for easily read and access to php config ar...
composer/ca-bundle                1.2.11  Lets you find a path to the system CA bundle, and includes a fallback to the Mozill...
danielstjules/stringy             3.1.0   A string manipulation library with multibyte support
daverandom/libdns                 v2.0.2  DNS protocol implementation written in pure PHP
doctrine/instantiator             1.4.0   A small, lightweight utility to instantiate objects in PHP without invoking their c...
doctrine/lexer                    1.2.1   PHP Doctrine Lexer parser library that can be used in Top-Down, Recursive Descent P...
egulias/email-validator           3.1.2   A library for validating emails against several RFCs
guzzlehttp/guzzle                 7.3.0   Guzzle is a PHP HTTP client library
guzzlehttp/promises               1.5.0   Guzzle promises library
guzzlehttp/psr7                   1.8.3   PSR-7 message implementation that also provides common utility methods
lcobucci/jwt                      3.4.6   A simple library to work with JSON Web Token and JSON Web Signature
league/flysystem                  1.1.5   Filesystem abstraction: Many filesystems, one API.
league/flysystem-memory           1.0.2   An in-memory adapter for Flysystem.
league/flysystem-sftp             1.0.22  Flysystem adapter for SFTP
league/mime-type-detection        1.8.0   Mime-type detection for Flysystem
monolog/monolog                   1.26.1  Sends your logs to files, sockets, inboxes, databases and various web services
mtdowling/jmespath.php            2.6.1   Declaratively specify how to extract elements from a JSON document
padraic/humbug_get_contents       1.1.2   Secure wrapper for accessing HTTPS resources with file_get_contents for PHP 5.3+
padraic/phar-updater              v1.0.6  A thing to make PHAR self-updating easy and secure.
phpdocumentor/reflection-common   2.2.0   Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock 5.2.2   With this component, a library can provide support for annotations via DocBlocks or...
phpdocumentor/type-resolver       1.5.1   A PSR-5 based resolver of Class names, Types and Structural Element Names
phpseclib/phpseclib               2.0.33  PHP Secure Communications Library - Pure-PHP implementations of RSA, AES, SSH2, SFT...
phpspec/prophecy                  1.14.0  Highly opinionated mocking framework for PHP 5.3+
psr/container                     1.1.1   Common Container Interface (PHP FIG PSR-11)
psr/http-client                   1.0.1   Common interface for HTTP clients
psr/http-message                  1.0.1   Common interface for HTTP messages
psr/log                           1.1.4   Common interface for logging libraries
ralouphie/getallheaders           3.0.3   A polyfill for getallheaders.
sebastian/comparator              3.0.3   Provides the functionality to compare PHP values for equality
sebastian/diff                    3.0.3   Diff implementation
sebastian/exporter                3.1.3   Provides the functionality to export PHP variables for visualization
sebastian/recursion-context       3.0.1   Provides functionality to recursively process PHP variables
swiftmailer/swiftmailer           v6.2.7  Swiftmailer, free feature-rich PHP mailer
symfony/config                    v5.3.4  Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                   v5.3.7  Eases the creation of beautiful and testable command line interfaces
symfony/dependency-injection      v5.3.8  Allows you to standardize and centralize the way objects are constructed in your ap...
symfony/deprecation-contracts     v2.4.0  A generic function and convention to trigger deprecation notices
symfony/filesystem                v5.3.4  Provides basic utilities for the filesystem
symfony/finder                    v5.3.7  Finds files and directories via an intuitive fluent interface
symfony/phpunit-bridge            v5.3.8  Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-ctype            v1.23.0 Symfony polyfill for ctype functions
symfony/polyfill-iconv            v1.23.0 Symfony polyfill for the Iconv extension
symfony/polyfill-intl-grapheme    v1.23.1 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-idn         v1.23.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer  v1.23.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring         v1.23.1 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72            v1.23.0 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73            v1.23.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80            v1.23.1 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/polyfill-php81            v1.23.0 Symfony polyfill backporting some PHP 8.1+ features to lower PHP versions
symfony/serializer                v5.3.8  Handles serializing and deserializing data structures, including object graphs, int...
symfony/service-contracts         v2.4.0  Generic abstractions related to writing services
symfony/string                    v5.3.7  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points ...
symfony/var-dumper                v5.3.8  Provides mechanisms for walking through any arbitrary PHP variable
symfony/yaml                      v5.3.6  Loads and dumps YAML files
webmozart/assert                  1.10.0  Assertions to validate method input/output with nice error messages.
webmozart/path-util               2.3.0   A robust cross-platform utility for normalizing, comparing and modifying file paths.
```

**Comment by SamMousa** [comment]
Are you using the PHAR when trying from master?

**Comment by aik099** [comment]
No. I'm using git clone from GitHub and then doing a `composer update` command.

**Comment by SamMousa** [comment]
Hmm, without a stack trace it'll be hard to track this down. Are you able to get a stack trace? Possibly by running the CLI in verbose / debug mode?

- Note I do not use the CLI so am not familiar with its quirks, if you get a stack trace we could look into what's wrong.

**Comment by aik099** [comment]
Here is the same command output, but with `-vvv` option:

```
Registering contact my@email.com...
Loading repository  
Creating Acme client  
Loading account key pair...  
Loading repository  
Loading repository  
Registering on the ACME server...
Account registered successfully!
Loading repository  

In Repository.php line 199:
                                                            
  [AcmePhp\Cli\Exception\AcmeCliException]                  
  Loading of domain my.domain.com distinguished name failed  
                                                            

Exception trace:
  at /path/to/acmephp/src/Cli/Repository/Repository.php:199
 AcmePhp\Cli\Repository\Repository->loadDomainDistinguishedName() at /path/to/acmephp/src/Cli/Command/RunCommand.php:217
 AcmePhp\Cli\Command\RunCommand->isUpToDate() at /path/to/acmephp/src/Cli/Command/RunCommand.php:87
 AcmePhp\Cli\Command\RunCommand->execute() at /path/to/acmephp/vendor/symfony/console/Command/Command.php:326
 Symfony\Component\Console\Command\Command->run() at /path/to/acmephp/vendor/symfony/console/Application.php:1078
 Symfony\Component\Console\Application->doRunCommand() at /path/to/acmephp/vendor/symfony/console/Application.php:324
 Symfony\Component\Console\Application->doRun() at /path/to/acmephp/vendor/symfony/console/Application.php:175
 Symfony\Component\Console\Application->run() at /path/to/acmephp/bin/acme:42

In Serializer.php line 220:
                                                                                                         
  [Symfony\Component\Serializer\Exception\NotNormalizableValueException]                                 
  Could not denormalize object of type "AcmePhp\Ssl\DistinguishedName", no supporting normalizer found.  
                                                                                                         

Exception trace:
  at /path/to/acmephp/vendor/symfony/serializer/Serializer.php:220
 Symfony\Component\Serializer\Serializer->denormalize() at /path/to/acmephp/vendor/symfony/serializer/Serializer.php:152
 Symfony\Component\Serializer\Serializer->deserialize() at /path/to/acmephp/src/Cli/Repository/Repository.php:197
 AcmePhp\Cli\Repository\Repository->loadDomainDistinguishedName() at /path/to/acmephp/src/Cli/Command/RunCommand.php:217
 AcmePhp\Cli\Command\RunCommand->isUpToDate() at /path/to/acmephp/src/Cli/Command/RunCommand.php:87
 AcmePhp\Cli\Command\RunCommand->execute() at /path/to/acmephp/vendor/symfony/console/Command/Command.php:326
 Symfony\Component\Console\Command\Command->run() at /path/to/acmephp/vendor/symfony/console/Application.php:1078
 Symfony\Component\Console\Application->doRunCommand() at /path/to/acmephp/vendor/symfony/console/Application.php:324
 Symfony\Component\Console\Application->doRun() at /path/to/acmephp/vendor/symfony/console/Application.php:175
 Symfony\Component\Console\Application->run() at /path/to/acmephp/bin/acme:42

run [-d|--delay DELAY] [--] <config>
```

Dependency versions:
```
adbario/php-dot-notation          2.5.0     PHP dot notation access to arrays
alibabacloud/cdn                  1.8.958   Alibaba Cloud Cdn SDK for PHP
alibabacloud/client               1.5.32    Alibaba Cloud Client for PHP - Use Alibaba Cloud in your PHP project
alibabacloud/wafopenapi           1.8.884   Alibaba Cloud WafOpenapi SDK for PHP
aws/aws-crt-php                   v1.2.6    AWS Common Runtime for PHP
aws/aws-sdk-php                   3.320.1   AWS SDK for PHP - Use Amazon Web Services in your PHP project
clagiordano/weblibs-configmanager v1.2.0    weblibs-configmanager is a tool library for easily read and access to php config ...
composer/semver                   3.4.2     Semver library that offers utilities, version constraint parsing and validation.
consolidation/self-update         2.2.0     Provides a self:update command for Symfony Console applications.
guzzlehttp/guzzle                 7.9.2     Guzzle is a PHP HTTP client library
guzzlehttp/promises               2.0.3     Guzzle promises library
guzzlehttp/psr7                   2.7.0     PSR-7 message implementation that also provides common utility methods
lcobucci/jwt                      5.3.0     A simple library to work with JSON Web Token and JSON Web Signature
league/flysystem                  3.28.0    File storage abstraction for PHP
league/flysystem-local            3.28.0    Local filesystem adapter for Flysystem.
league/flysystem-memory           3.28.0    In-memory filesystem adapter for Flysystem.
league/flysystem-sftp-v3          3.28.0    SFTP filesystem adapter for Flysystem.
league/mime-type-detection        1.15.0    Mime-type detection for Flysystem
monolog/monolog                   2.9.3     Sends your logs to files, sockets, inboxes, databases and various web services
mtdowling/jmespath.php            2.7.0     Declaratively specify how to extract elements from a JSON document
paragonie/constant_time_encoding  v3.0.0    Constant-time Implementations of RFC 4648 Encoding (Base-64, Base-32, Base-16)
paragonie/random_compat           v9.99.100 PHP 5.x polyfill for random_bytes() and random_int() from PHP 7
phpseclib/phpseclib               3.0.41    PHP Secure Communications Library - Pure-PHP implementations of RSA, AES, SSH2, S...
phpstan/phpstan                   1.11.10   PHPStan - PHP Static Analysis Tool
psr/clock                         1.0.0     Common interface for reading the clock.
psr/container                     1.1.2     Common Container Interface (PHP FIG PSR-11)
psr/http-client                   1.0.3     Common interface for HTTP clients
psr/http-factory                  1.1.0     PSR-17: Common interfaces for PSR-7 HTTP message factories
psr/http-message                  1.1       Common interface for HTTP messages
psr/log                           1.1.4     Common interface for logging libraries
ralouphie/getallheaders           3.0.3     A polyfill for getallheaders.
symfony/config                    v6.4.8    Helps you find, load, combine, autofill and validate configuration values of any ...
symfony/console                   v6.4.10   Eases the creation of beautiful and testable command line interfaces
symfony/dependency-injection      v6.4.10   Allows you to standardize and centralize the way objects are constructed in your ...
symfony/deprecation-contracts     v3.5.0    A generic function and convention to trigger deprecation notices
symfony/filesystem                v6.4.9    Provides basic utilities for the filesystem
symfony/finder                    v6.4.10   Finds files and directories via an intuitive fluent interface
symfony/phpunit-bridge            v6.4.10   Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-ctype            v1.30.0   Symfony polyfill for ctype functions
symfony/polyfill-intl-grapheme    v1.30.0   Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-normalizer  v1.30.0   Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring         v1.30.0   Symfony polyfill for the Mbstring extension
symfony/property-access           v6.4.8    Provides functions to read and write from/to an object or array using a simple st...
symfony/property-info             v7.1.3    Extracts information about PHP class' properties using metadata of popular sources
symfony/serializer                v6.4.10   Handles serializing and deserializing data structures, including object graphs, i...
symfony/service-contracts         v3.5.0    Generic abstractions related to writing services
symfony/string                    v7.1.3    Provides an object-oriented API to strings and deals with bytes, UTF-8 code point...
symfony/type-info                 v7.1.1    Extracts PHP types information.
symfony/var-dumper                v6.4.10   Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter              v7.1.2    Allows exporting any serializable PHP data structure to plain PHP code
symfony/yaml                      v6.4.8    Loads and dumps YAML files
webmozart/assert                  1.11.0    Assertions to validate method input/output with nice error messages.
```

**Comment by aik099** [comment]
I've debugged a bit (working code VS non-working code) and the problem comes from changes in the `\Symfony\Component\Serializer\Serializer::getDenormalizer` method. It uses `$this->denormalizerCache`, but doesn't populate it as in older Symfony Serializer versions when normalizers have `getSupportedTypes` defined (which both of them have).

**Comment by SamMousa** [comment]
Could you make a PR containing a (failing) test case? That's the most important thing for easy reproduction. If it contains a fix as well that'd be awesome of course ;-)

**Comment by aik099** [comment]
The problem is inside Symfony Serializer (so the PR should go there), when you give it custom class to deserialize into (e.g. `AcmePhp\Ssl\DistinguishedName`) it no longer uses `Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer` to deserialize it and throws the exception.

Code: https://github.com/symfony/serializer/blob/6.4/Serializer.php#L404 .

I don't understand the logic behind this and therefore I can't send a PR.

**Comment by lyrixx** [comment]
can you provide a reproducer?

**Comment by aik099** [comment]
> can you provide a reproducer?

@lyrixx , here are the instructions to make the error happen (not sure which steps are relevant, that's why I've listed all I have):

1. create this domain config file:

```yaml
contact_email: my@email.com

defaults:
  distinguished_name:
      country: MN
      locality: Erdenet
      organization_name: OrgName
  solver: http

certificates:
  - domain: main.my.domain
    subject_alternative_names:
      - '*.main.my.domain'
      - 'sub.my.ninja'
      - '*.sub1.main.my.domain'
      - '*.sub2.main.my.domain'
```
2. create acmephp config file in the `~/.acmephp/acmephp.conf`:
```
#
# acmephp.conf - Configuration file for Acme PHP client.
#
# See https://github.com/acmephp/acmephp for more informations
#

###################################################################
# Storage
#
# Configure here where and how you want to save your certificates
# and SSL keys.
#

storage:

    #
    # By default, Acme PHP will create a backup of every file
    # before any modification. You can disable this mechanism here.
    #
    enable_backup: true

    #
    # Actions to execute right after the generation of a file (key, CSR or certificate).
    # Actions are executed in the order provided in configuration.
    #
    # See the documentation to learn how to create your own action.
    #
    post_generate: ~

        #
        # Build a file structure suited for nginx-proxy (https://github.com/jwilder/nginx-proxy)
        #
        # - action: build_nginxproxy

        #
        # Mirror your local storage directory on a FTP host
        #
        # - action: push_ftp
        #   root: /acmephp
        #   host: ftp.example.com
        #   username: username
        #   password: password
        #   # port: 21
        #   # passive: true
        #   # ssl: true
        #   # timeout: 30

        #
        # Mirror your local storage directory on a SFTP host
        #
        # - action: push_sftp
        #   root: /acmephp
        #   host: sftp.example.com
        #   username: username
        #   password: password
        #   # port: 21
        #   # private_key: path/to/or/contents/of/privatekey
        #   # timeout: 30


###################################################################
# Monitoring
#
# This section let you configure a simple monitoring mechanism that
# will warn you if an error occurs during a CRON job.
#

monitoring: ~   # Monitoring is disabled by default

# You can enabled it by configuring at least one alert handler.
# You can change the default handler level to decide when to be alerted
# (only when an error occurs or every time the CRON is started).
#monitoring:
#    email:   # Only SMTP(S) support for now
#        to: galopintitouan@gmail.com
#        host: smtp.example.com
#        # port: 25
#        # username: ~
#        # password: ~
#        # encryption: ~
#        # subject: An error occured during Acme PHP CRON renewal
#        # level: error     # By default, only on error for email handler
#
#    slack:
#        token: your_token
#        channel: general   # Channel name without hashtag
#        # username: Acme PHP
#        # level: info      # By default, on every CRON for slack handler
```
3. have `master` branch of this repo checked out somewhere with all the latest composer dependencies installed
4. run the `/path/to/cloned/repo/bin/acme run /path/to/domain/config/file.yml` command
5. everything would work
6. run the same command 
7. the above command would fail to read data from the previously generated SSL certificates

**Comment by lyrixx** [comment]
Thanks for the reproducer. I managed to reproduce the issue, and fix it #312
